### PR TITLE
Fix bug 27490: Can't change customer group when placing an admin order

### PR DESCRIPTION
**Pull request by tna274**
### Description (*)
Fix bug 27490: Can't change customer group when placing an admin order

### Fixed Issues (if relevant)

1. magento/magento2#27490: Fix bug 27490: Can't change customer group when placing an admin order

### Manual testing scenarios (*)

1. In Admin System > Configurations > Customers > Customer Configurations > Create New Account Options > Enable Automatic Assignment to Customer Group >> Select YES
2. In Admin Sales > Orders.
3. Select Create New Order
4. Select Create New Customer
5. Under Account Information, change the group.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: app/code/Magento/Quote/Observer/Frontend/Quote/Address/CollectTotalsObserver.php [hunk 1]
```diff
@@ -78,6 +89,7 @@ public function __construct(
         $this->groupManagement = $groupManagement;
         $this->addressRepository = $addressRepository;
         $this->customerSession = $customerSession;
+        $this->state = $state ?: ObjectManager::getInstance()->get(State::class);
```
Comment (Stepa4man) [inline app/code/Magento/Quote/Observer/Frontend/Quote/Address/CollectTotalsObserver.php]: Once this PR will be delivered in a major release of Magento we can avoid using BC way of adding new dependencies to the constructor of non **@api** classes.

File: app/code/Magento/Quote/Observer/Frontend/Quote/Address/CollectTotalsObserver.php [hunk 2]
```diff
@@ -131,7 +143,12 @@ public function execute(\Magento\Framework\Event\Observer $observer)
             );
         }
 
-        if ($groupId !== null) {
+        // Do not update customer group as not logged in when doing process in admin area
+        if ($groupId !== null
+            && !(
+                $this->state->getAreaCode() == Area::AREA_ADMINHTML
```
Comment (Stepa4man) [inline app/code/Magento/Quote/Observer/Frontend/Quote/Address/CollectTotalsObserver.php]: Please use a strict comparison operator `===` and add `declare(strict_types=1);`

**Review by Stepa4man — CHANGES_REQUESTED** [review]
Hi @tna274 
Thanks for your contribution.

Please change your code according to [Magento technical guideline ](https://devdocs.magento.com/guides/v2.3/coding-standards/technical-guidelines.html#1-basic-programming-principles)

Also please consider refactoring this class.
Instead of introducing new code complexity try to separate implementation for adminhtml and frontend introducing 2 observers and declare them at respective levels via events.xml
To avoid code duplication please use decomposition.

**Comment by tna274** [comment]
Hi @Stepa4man 
I have updated my pull request

**Comment by engcom-Charlie** [comment]
Hi, @tna274 can you please look through failed tests and merge conflicts?

**Comment by engcom-Kilo** [comment]
@magento run all tests

**Comment by engcom-Kilo** [comment]
Working with MFTF test.

**Comment by engcom-Kilo** [comment]
@magento run Magento Health Index

**Comment by engcom-Kilo** [comment]
@magento run Static Tests

**Comment by engcom-Kilo** [comment]
Added MFTF test.

**Comment by engcom-Kilo** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Kilo** [comment]
@magento run all tests

File: app/code/Magento/Quote/Observer/Frontend/Quote/Address/CollectTotalsObserver.php [hunk 3]
```diff
@@ -131,12 +131,23 @@ public function execute(\Magento\Framework\Event\Observer $observer)
...
+
+    /**
+     * Conditions to change customer group
+     *
+     * @param int|null $groupId
+     * @return bool
+     */
+    protected function assignCustomerGroupConditions($groupId)
+    {
+        return $groupId !== null ? true : false;
```
Comment (Stepa4man) [inline app/code/Magento/Quote/Observer/Frontend/Quote/Address/CollectTotalsObserver.php:151]: Why do you need it here? Why don't you just write `return $groupId !== null`?

File: app/code/Magento/Quote/Observer/Backend/Quote/Address/CollectTotalsObserver.php [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    {
+        if ($groupId !== null
+            && !(
+                $this->state->getAreaCode() == Area::AREA_ADMINHTML
+                && $groupId == $this->groupManagement->getNotLoggedInGroup()->getId()
+            )) {
+            return true;
+        }
+
+        return false;
```
Comment (Stepa4man) [inline app/code/Magento/Quote/Observer/Backend/Quote/Address/CollectTotalsObserver.php:79]: Please put those long difficult conditions to separate private methods with proper naming and shorten this function to just return statement:
`
return $groupId !== null && !($this->isAdminArea() && $this->isNotLoggedInGroup($groupId));
`
At least like this.

You can think even more to make this code more readable.

File: app/code/Magento/Quote/Test/Unit/Observer/Backend/Quote/Address/CollectTotalsObserverTest.php [hunk 5]
```diff
@@ -0,0 +1,488 @@
...
+ *
+ * @SuppressWarnings(PHPMD.CouplingBetweenObjects)
+ * @SuppressWarnings(PHPMD.TooManyFields)
+ */
+class CollectTotalsObserverTest extends \PHPUnit\Framework\TestCase
+{
+    /**
+     * @var \Magento\Quote\Observer\Backend\Quote\Address\CollectTotalsObserver
+     */
+    protected $model;
```
Comment (Stepa4man) [inline app/code/Magento/Quote/Test/Unit/Observer/Backend/Quote/Address/CollectTotalsObserverTest.php:24]: Please replace all protected properties with private.

File: app/code/Magento/Quote/Test/Unit/Observer/Backend/Quote/Address/CollectTotalsObserverTest.php [hunk 6]
```diff
@@ -0,0 +1,488 @@
...
+
+    /**
+     * @var MockObject
```
Comment (Stepa4man) [inline app/code/Magento/Quote/Test/Unit/Observer/Backend/Quote/Address/CollectTotalsObserverTest.php:27]: Please add the proper type annotation to all class properties

File: app/code/Magento/Quote/Test/Unit/Observer/Backend/Quote/Address/CollectTotalsObserverTest.php [hunk 7]
```diff
@@ -0,0 +1,488 @@
...
+     * @var State|MockObject
+     */
+    protected $stateMock;
+
+    /**
+     * @SuppressWarnings(PHPMD.ExcessiveMethodLength)
+     */
+    protected function setUp()
+    {
+        $this->objectManager = new \Magento\Framework\TestFramework\Unit\Helper\ObjectManager($this);
```
Comment (Stepa4man) [inline app/code/Magento/Quote/Test/Unit/Observer/Backend/Quote/Address/CollectTotalsObserverTest.php:106]: Please don't use ObjectManager helper, use `new` statement instead.

File: app/code/Magento/Quote/Test/Unit/Observer/Backend/Quote/Address/CollectTotalsObserverTest.php [hunk 8]
```diff
@@ -0,0 +1,488 @@
...
+        $this->storeId = 1;
+        $this->customerMock = $this->getMockForAbstractClass(
+            \Magento\Customer\Api\Data\CustomerInterface::class,
```
Comment (Stepa4man) [inline app/code/Magento/Quote/Test/Unit/Observer/Backend/Quote/Address/CollectTotalsObserverTest.php:109]: Please import all class names.

File: app/code/Magento/Quote/Observer/Backend/Quote/Address/CollectTotalsObserver.php [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+     *
+     * @param int|null $groupId
+     * @return bool
+     */
+    protected function assignCustomerGroupConditions($groupId)
+    {
+        if ($groupId !== null
+            && !(
+                $this->state->getAreaCode() == Area::AREA_ADMINHTML
+                && $groupId == $this->groupManagement->getNotLoggedInGroup()->getId()
```
Comment (Stepa4man) [inline app/code/Magento/Quote/Observer/Backend/Quote/Address/CollectTotalsObserver.php:74]: Please use strict comparison operator

File: app/code/Magento/Quote/Test/Unit/Observer/Backend/Quote/Address/CollectTotalsObserverTest.php [hunk 10]
```diff
@@ -0,0 +1,488 @@
+<?php
```
Comment (Stepa4man) [inline app/code/Magento/Quote/Test/Unit/Observer/Backend/Quote/Address/CollectTotalsObserverTest.php:1]: Please declare strict types for every php file

**Review by Stepa4man — CHANGES_REQUESTED** [review]
Hi @tna274 
Thanks for your contribution.

Please take a look at my comments.

**Comment by sidolov** [comment]
@tna274  I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!

**Pull request closed without merging**